webapp: Enable adding, removing, and listing hybrid connections and vnet integrations on web apps

singlessis opened the pull request:
Fixes #8453 
---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [x] The PR has modified HISTORY.rst describing any customer-facing, functional changes. Note that this does not include changes only to help content. (see [Modifying change log](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules#modify-change-log)).

- [x] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

msftclas commented:
[![CLA assistant check](https://cla.opensource.microsoft.com/pull/badge/not_signed)](https://cla.opensource.microsoft.com/Azure/azure-cli?pullRequest=8251) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.opensource.microsoft.com/Azure/azure-cli?pullRequest=8251) before we can accept your contribution.<br/><br/>:x: singlessis [sign now](https://cla.opensource.microsoft.com/Azure/azure-cli?pullRequest=8251)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.microsoft.com/check/Azure/azure-cli?pullRequest=8251) it.</sub>

tjprescott commented on src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/_help.py:
```diff
@@ -974,3 +974,71 @@
     short-summary: show allowed origins
 """
+
+helps['webapp hybridconnection list'] = """
```
hybridconnection should be hyphenated: `hybrid-connection`.

tjprescott commented on src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/_help.py:
```diff
@@ -974,3 +974,71 @@
+            az webapp vnetintegration list -g <myRG> -n <myAppName>
+"""
+
+helps['webapp vnetintegration set'] = """
```
It looks like these commands should be `add/remove` instead of `set/remove`.

tjprescott commented on src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/_help.py:
```diff
@@ -974,3 +974,71 @@
+    examples:
+        - name: list the hyrbid connections
+          text: >
+            az webapp vnetintegration list -g <myRG> -n <myAppName> --vnet-resource-group <vnetRG> --vnet-name <vnetName>
```
vnetintegration should be `vnet-integration`.  Placeholders should use `{thisSyntax}` not `<thisSyntax>`. `vnet-resource-group` violates the CLI command guidelines:
https://github.com/Azure/azure-cli/blob/dev/doc/authoring_command_modules/authoring_commands.md#supporting-name-or-id-parameters

panchagnula commented:
@singlessis can you update the title to be specific & descriptive of the features we are trying to support here? Also please sync & merge to upstream dev branch so that the CI build passes, then a next pass on the PR can be done. I also recommend creating a gibhub issue(/s) describing the feature & commands are you are adding support for here & linking them to this PR. Thank you.

panchagnula commented:
@singlessis I updated the title for the PR , please update this further if needed. Also, should we support this for function app as well? if so this is an easy change, look at some of the existing commands which should map the method to function app as well.

singlessis commented:
@tjprescott @yugangw-msft is there anything else I need to do before this is approved? My pm wants to demo this soon, so it would be great to get it out. Thank you!

panchagnula commented:
@singlessis is this ready for review? Or do we need the SDK changes before we get this in? If the SDK issues are blocking this PR - we should start an email with Yugang & his team to help us with the specific changes we need, especially since our swagger is not generating the SDK changes as required.

panchagnula commented:
//cc: @ccompy

singlessis commented:
@yugangw-msft 
I am experiencing an issue with my vnet-integration functions. The vnet objects contain a byte array, which is not json serializable. I added a custom serializing function to my python code on my local computer, which allowed all my functions to work. However, this can't really be propagated, have you encountered anything like this, that required custom serialization? thank you!

This is the function that I added to C:\Users\soingles\AppData\Local\Programs\Python>Python37-32\Lib\json\encoder.py

        if isinstance(o, bytearray):     
            strRep = base64.b64encode(o)
            return strRep
        else:
            raise TypeError(repr(python_object) + ' is not JSON serializable')

panchagnula commented:
@yugangw-msft or @tjprescott can you comment on Sofi's comment about having to update the encoder.py to get this working, we looked at the Swagger from our end & did not see anything obvious that could cause a need for updating the encoder.py. Thank you.

yugangw-msft commented:
@panchagnula, what is the API involved? Can you point out the related swagger spec or SDK method?

singlessis commented:
@yugangw-msft I am calling the method list_vnet_connections, in web_apps_operations.py. When the vnet that is to be returned is a non-swift vnet, I run into that error

yugangw-msft commented:
I assume the relevant field is the `cert_blob`. Please clarify what did you mean exactly, particularly `this can't really be propagated`. Does the SDK API invoke fail? If yes, what is the exception, and stack trace?

singlessis commented:
When I originally tried to call the list_vnet function, I received this error: 
Object of type bytearray is not JSON serializable
Traceback (most recent call last):
  File "C:\Users\soingles\azure-cli\env\lib\site-packages\knack\cli.py", line 201, in invoke
    self.output.out(cmd_result, formatter=formatter, out_file=out_file)
  File "C:\Users\soingles\azure-cli\env\lib\site-packages\knack\output.py", line 132, in out
    output = formatter(obj)
  File "C:\Users\soingles\azure-cli\env\lib\site-packages\knack\output.py", line 40, in format_json
    separators=(',', ': ')) + '\n'
  File "C:\Users\soingles\AppData\Local\Programs\Python\Python37-32\lib\json\__init__.py", line 238, in dumps
    **kw).encode(obj)
  File "C:\Users\soingles\AppData\Local\Programs\Python\Python37-32\lib\json\encoder.py", line 201, in encode
    chunks = list(chunks)
  File "C:\Users\soingles\AppData\Local\Programs\Python\Python37-32\lib\json\encoder.py", line 429, in _iterencode
    yield from _iterencode_list(o, _current_indent_level)
  File "C:\Users\soingles\AppData\Local\Programs\Python\Python37-32\lib\json\encoder.py", line 325, in _iterencode_list
    yield from chunks
  File "C:\Users\soingles\AppData\Local\Programs\Python\Python37-32\lib\json\encoder.py", line 405, in _iterencode_dict
    yield from chunks
  File "C:\Users\soingles\AppData\Local\Programs\Python\Python37-32\lib\json\encoder.py", line 438, in _iterencode
    o = _default(o)
  File "C:\Users\soingles\azure-cli\env\lib\site-packages\knack\output.py", line 32, in default
    return json.JSONEncoder.default(self, o)
  File "C:\Users\soingles\AppData\Local\Programs\Python\Python37-32\lib\json\encoder.py", line 179, in default
    raise TypeError(f'Object of type {o.__class__.__name__} '
TypeError: Object of type bytearray is not JSON serializable

In order to get around this, I added code in the encoder.py function,  in order to deal with the byte array. Since this is not a file that is included with the cli package, i was not sure how that would be passed on to the users, which is what I meant by propagation.

yugangw-msft commented:
This happens at the command output stage. I suggest you decode it to string in you command code before send it to output

singlessis commented:
Do you mean in the web_apps_operation file itself?

yugangw-msft commented:
no, i mean in the custom.py right after you invoke the vnet method from the `web_apps_operation`

singlessis commented:
When I receive this error, I never successfully return to custom.py, the whole thing fails inside web_apps.

yugangw-msft commented:
@singlessis @panchagnula  is this PR ready for review? If yes, it has to be rebased to eliminate irrelevant change.

panchagnula commented:
The VNET part of it is not ready@yugangw-msft , we still need the swagger update like you pointed out in the debugging session @singlessis you have two options
1. Keep this PR restricted to Hybrid Connections only & get it in this Sprint
2. If we want both features to be released together, then move this to the next sprint.

yugangw-msft commented:
I am removing the milestone tag and feel free to add back when the code is ready

tjprescott commented:
This PR has been open almost 3 months and the last commits were over a month ago. Closing. Please reopen when you are ready for it to be reviewed.
